### PR TITLE
Initialization of the BSS is done by the boot hart.

### LINKED ISSUE #151: Multicore-support for BSS initialization
In the current implementation all threads are zeroing out the BSS, which can lead to race condition and loss of data.

https://github.com/CharlyCst/miralis/blob/332ed085096d624f59beb2fb3afa105b579b043b/src/arch/metal.rs#L771-L781

To allow for correct execution on multicore platforms we need to ensure that:
- A single hart zeroes-out the BSS section
- No hart exits the assembly boot code before the BSS section is zeroed out.

In practice that means that we need to pick a boot hart and have it zero-out all of the BSS while the other harts are waiting.
We already have a concept of boot hart, so we can re-use it to choose which core should boot first rather than doing a race.
For all other harts we need to setup a synchronization barrier. The easiest option is to wait on an atomic, the fancy version would be to do a `wfi` but that makes the implementation drastically more complication so I would suggest to stick to the basic for now.

### PR DESCRIPTION
**Pull request by NoeTerrier**
All other hart wait for the primary hart to initialize the BSS before jumping to main. This avoids race condition on the BSS section, which should be zero.

This is maybe missing a test. How could it be tested ?

Close #151

### CODE REVIEW

File: src/arch/metal.rs
Comment (CharlyCst) [inline src/arch/metal.rs:841]: Rust does not give layout guarantees for booleans. So this code might break when upgrading the compiler if the layout changes for some reason. Also big/little endian change the conversions between boolean and integers.

So you should pick a well defined integer size and initialize it with a proper layout. Then be sure that your load matches the size of the integer type you picked.

File: src/arch/metal.rs
Comment (CharlyCst) [inline src/arch/metal.rs:780]: You load the address of the boolean _after_ the jump of the non-boot harts. So the non-boot harts are truing to read from `t3` but `t3` might contain some invalid address.

You should load it before to ensure all threads have the address of the variable in `t3`.

File: src/arch/metal.rs
Comment (CharlyCst) [inline src/arch/metal.rs:800]: Here you need an atomic too, probably with acquire semantic (you should always pair a release with an acquire to guarantee ordering, at least on other architecture I know about).

Comment (CharlyCst) [reply src/arch/metal.rs:800]: Actually I realize RISC-V might not call those atomics, but maybe just loads with acquire semantic?

**Review by CharlyCst — CHANGES_REQUESTED** [review]
Thanks for the changes!

I think there are some issued in the boot assembly, maybe that's what causing issues later on?

File: src/arch/metal.rs
Comment (NoeTerrier) [reply src/arch/metal.rs:841]: Has `usize` a well-defined integer size?

Comment (CharlyCst) [reply src/arch/metal.rs:841]: It is platform dependent (32 or 64 bits), but yes the layout is specified ^^

File: src/arch/metal.rs
Comment (NoeTerrier) [reply src/arch/metal.rs:800]: I can replace `amoswap.w.rl` by `amoswap.w.aqrl`. This means that the operation will be done in a sequential order in respect to the BSS zeroing. It will be visible by other hart in a sequential order too.

We do not need the other harts to acquire a lock, but only to pull the boolean until it is set to 0 by the boot hart.
The boolean is set to 1 by default, and cleared only by hart 0. We need to ensure that the BSS zeroing is done and visible before the boolean clear. So sequential ordering of the boolean clear should work, right?

Spec:

> To help implement multiprocessor synchronization, the AMOs optionally provide release consis-
tency semantics. If the aq bit is set, then no later memory operations in this RISC-V hart can be
observed to take place before the AMO. Conversely, if the rl bit is set, then other RISC-V harts will
not observe the AMO before memory accesses preceding the AMO in this RISC-V hart. Setting
both the aq and the rl bit on an AMO makes the sequence sequentially consistent, meaning that it
cannot be reordered with earlier or later memory operations from the same hart.

Comment (CharlyCst) [reply src/arch/metal.rs:798]: The acquire and release semantic are local to a hart, by themselves they do not change what other harts can see.

Here the release semantic means that every operations before the AMO will be ordered before the AMO, while the acquire means that no operation that happens after the AMO will be ordered before the AMO. But you have to think of that as the other in which the hart will submit the changes to the memory system (caches, etc...).

To see a consistent ordering between two harts we need two things:
1. The hart doing the writes needs to decide on its own ordering (the order in which it submits changes to the memory subsystem).
2. The hart doing the reads needs to check if there were updated on the value it reads, so it needs to ask the memory subsystem if anything changed and in which order.

You are doing (1), but not (2), because you don't use a fence or atomic operation to read the value from memory.
Feel free to ask if anything is unclear :)

**Review by CharlyCst — COMMENTED** [review]
See the discussion on atomics, feel free to ask if you want to discuss that particular point :)

File: src/main.rs
Comment (CharlyCst) [inline src/main.rs:65]: We might now want to delete this in this PR, as the rest of the code is not ready for multi-core I assume.

I'll handle it before merging :)

**Review by CharlyCst — APPROVED** [review]
Thanks, looks good. I'll merge the PR when I get some time, thanks for the changes!

**Pull request merged**
